File: Construct binary search tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        return bintreeform(preorder,0,preorder.size()-1);
    }
    TreeNode* bintreeform(vector<int>& preorder,int l,int r)
    {
        if(l>r)
            return NULL;
        TreeNode* root=new TreeNode(preorder[l]);
        if(l==r)
            return root;
        int midindx=l+1;
        while(midindx<=r && preorder[midindx]<preorder[l])
            midindx++;
        root->left=bintreeform(preorder,l+1,midindx-1);
        root->right=bintreeform(preorder,midindx,r);
        return root;
        
    }
};

File: Construct binary search tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode Constructbst(TreeNode root,int data)
    {
        if(root==null)
        {
            root=new TreeNode(data);
            return root;
        }
        else if(root.val>data)
        {
            root.left=Constructbst(root.left,data);
        }
        else
        {
            root.right=Constructbst(root.right,data);
        }
        return root;
    }
    public TreeNode bstFromPreorder(int[] preorder) {
        int n=preorder.length;
        if(n==0)
            return null;
        TreeNode root=new TreeNode(preorder[0]);
        for(int i=1;i<n;i++)
        {
            root=Constructbst(root,preorder[i]);
        }
        return root;
    }
}
